lax.custom_linear_solve primitive

shoyer opened the pull request:
A follow-on to `lax.root` from https://github.com/google/jax/pull/1339. See the tests for a simple example of defining gradients for an iterative linear solver.

Eventually, I expect to use this for implementing iterative solvers (e.g., GMRES, CG) from `scipy.sparse.linalg`, which in turn we will use for the linear solves needed for gradients in `scipy.optimize.root`.

shoyer committed:
WIP: linear solvers

shoyer committed:
Merge branch 'solvers' into linear-solvers

shoyer committed:
Draft of lax.linear_solve

shoyer committed:
Merge branch 'master' into linear-solvers

shoyer committed:
Refactor pytree munging inside lax.root.

The primitive's implementation and JVP rules are now 100% pytree free.

shoyer committed:
Merge branch 'refactor-root' into linear-solvers

shoyer committed:
Fixup linear_solve

shoyer committed:
Linearize multiple times in _root_jvp to avoid zeros

shoyer committed:
Merge branch 'refactor-root' into linear-solvers

shoyer committed:
fix deftraced

shoyer committed:
add a symmetric argument

shoyer committed:
Fixup float64; add a test for symmetric/non-symmetric

shoyer committed:
test zeros in linear_solve_jvp

dougalm commented on jax/lax/lax_control_flow.py:
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
+  """Differentiably solve the linear map matvec(x)=b for x.
+
+  Required invariant:
+      x = solve(matvec, b)
```
`solve` should be `forward_solve`. Perhaps also describe what `(co)tangent_solve` is supposed to be and what invariants it observes?

shoyer replied:
Done, I wrote a full docstring.

dougalm commented on jax/lax/lax_control_flow.py:
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
+      _tangent_linear_map, core.jaxpr_as_fun(jaxpr), params, params_dot
+  )
+
+  if all(tangent is ad_util.zero for tangent in params_dot):
```
What happens if `params_dot` or `b_dot` are a mixture of zero and not-zero? Seems like it might blow up.

shoyer replied:
This should be handled now. I'm not entirely happy with the test coverage, but struggled to figure out how to come up with a test case that could involve a mixture of zeros/non-zeros -- I think I would need a three argument primitive?

dougalm commented on jax/lax/lax_control_flow.py:
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
+      error = matvec(x) - b
+      assert all(error == 0)
+  """
+  if tangent_solve is None:
```
Are these two cases exercised by the tests?

shoyer replied:
It is now :)

shoyer committed:
Revisions per review

shoyer committed:
Merge branch 'refactor-root' into linear-solvers

shoyer committed:
Adjust signature of linear_solve

shoyer committed:
Merge branch 'master' into linear-solvers

shoyer committed:
restore botched test

shoyer committed:
variable names

jekbradbury commented on jax/lax/lax_control_flow.py:
```diff
@@ -974,14 +977,141 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+xla.initial_style_translations[linear_solve_p] = xla.lower_fun(
+    _linear_solve_impl, initial_style=True)
+ad.primitive_transposes[linear_solve_p] = _linear_solve_transpose_rule
+batching.deftraced(linear_solve_p)
```
Do you have a sense of whether this will run into #1249?

shoyer replied:
Yeah, I was wondering the same thing...

shoyer replied:
Yeah, I was wondering the same thing...

shoyer replied:
Indeed, this most definitely suffers from the same issue. It appears we will need to write custom batching rules for `root` and `linear_solve`. I think this would be pretty doable, but now I'm leaving this as a "TODO".

shoyer committed:
jaxprize

shoyer committed:
spelling

shoyer committed:
Use np.dot instead of @

shoyer committed:
linear_solve docstring, more tests

shoyer committed:
Disable batching for root and linear_solve

shoyer committed:
Fix linear_solve tests

shoyer committed:
remove unused imports

shoyer commented:
Please take another look. I rewrote everything so `solve`/`transpose_solve` are now evaluated to JAXprs before calling the primitive, so they work on closures. Also added a bunch more tests/docs/validation.

gehring commented on jax/lax/lax_control_flow.py:
```diff
@@ -974,14 +976,194 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+    solve: higher level function that solves for solution to the linear
+      equation, i.e., ``matvec(solve(matvec, x)) == x`` for all ``x`` of the
+      same form as ``b``. This function need not be differenatiable.
+    transpose_solve: higher level function for solving the transpose linear
```
Is `transpose_solve` used when no derivatives are ever computed, i.e., just used as a regular solver? I'm just thinking that it would be good for requirement for `transpose_solve` should be as generous as possible.

Though this could be overly cautious on my part. Is there a case where someone would use `linear_solve` and not need derivatives given that the user already has to provide a `solve` function?

shoyer replied:
`transpose_solve` only gets used if a VJP needs to be calculated, e.g., for calculating backwards gradients with `jax.grad`. It's certainly possible to use this function for only the JVP rule, though I think this would be a little unusual.

I'll see if I can delay the error from not defining `transpose_solve` until the VJP rule without making things much more complex.

gehring replied:
Okay, that's what I thought! I agree that it is not worth changing it if it would bring in a lot of additional complexity. This can always be looked at if this becomes an actual limitation for someone rather than just a theoretical issue. Though if it's an easy fix, might as well do it now!

shoyer replied:
Done

gehring commented on jax/lax/lax_control_flow.py:
```diff
@@ -974,14 +976,194 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+                             tree_unflatten(tree, expected_shapes)))
+
+
+def linear_solve(matvec, b, solve, transpose_solve=None, symmetric=False):
```
Does it make sense to call it `linear_solve` if the user has to provide a function to solve the linear system? The purpose of this function is clear to me, but I'm trying to put myself in a new user's shoes.

A new user might not be familiar with the subtleties of automatic differentiation. This is similar to our `lax.root` discussion. It might be good to make sure it is clear that the purpose is to define "appropriate" derivatives. I think that something with `def*` prefix, ala `defjvp`, makes it clear but the docs alone could possibly do the trick. I'm afraid that a new user would read `linear_solve` and just assume to know what it does. An "unusual" name (like something with a `def*` prefix) would avoid users assuming they know what the function does just from the name.
WDYT?

shoyer replied:
I agree that an unusual name would be a good idea here. I'm leaning towards `differentiable_linear_solve` per our discussion over in https://github.com/google/jax/issues/1448. (I'll also definitely add clearer docs explaining why this is useful!)

shoyer replied:
Maybe `custom_linear_solve`?

gehring replied:
Our discussion is now a little split but I'm on board!

> I'll also definitely add clearer docs explaining why this is useful!

Great! Though, there is no need to over do it here, imo. 2-3 brief sentences should be enough for people to know what to look up if they don't understand. More detailed docs might deserve a standalone implicit diff page similar to async dispatch and co, that explains why backprop through all the things isn't the only solution.

shoyer committed:
Rename to custom_linear_solve

gehring commented on jax/lax/lax_control_flow.py:
```diff
@@ -974,14 +976,201 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+  Required invariant:
+      x = solve(matvec, b)  # solve the linear equation
+      assert matvec(x) == b  # not checked
+
```
> Maybe `custom_linear_solve`?

I like it! It's reminiscent of `jax.custom_gradient` and should cue users to pay attention to the docs. Also, the additional docs look great!

shoyer committed:
WIP: refactor

shoyer committed:
fixup test for lazy transpose_solve error

shoyer commented:
This is ready for a final (?) review.

dougalm commented on jax/lax/lax_control_flow.py:
```diff
@@ -974,14 +976,209 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+  """Transpose a linear function."""
+  # TODO(shoyer): can we use something more direct than the vjp machinery?
+  # It's particularly awkward that we need the second argument to give
+  # particular values of the primals, which are entirely arbitrary.
```
Good idea. It might be a new transformation, but one that shares a lot of machinery with the existing jvp/transpose.
